File: backend/app/schemas/incident.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


PageGroup = Literal["responder", "dispatchers"]


# Keep these request and response shapes aligned with contracts/openapi.json.
class ResponseRecordCreate(BaseModel):
    status: str
    source: str
    rank: int = 1


class ResponseRecordRead(BaseModel):
    user_public_id: str
    status: str
    rank: int
    updated: datetime


class IncidentRead(BaseModel):
    public_id: str
    title: str
    team_public_id: str | None = None
    location: str
    created: datetime
    notes: str
    active: bool
    responses: list[ResponseRecordRead]


class IncidentCreate(BaseModel):
    title: str
    team_public_id: str
    location: str
    notes: str
    active: bool = True


class IncidentUpdate(BaseModel):
    title: str
    location: str
    notes: str
    active: bool
    # Page-group selector for the optional incident.paged event that follows
    # incident.updated. Defaults to "responder" (page everyone) per
    # docs/page-logic.md. `null` skips paging entirely.
    page_group: PageGroup | None = "responder"

File: UserInterface/responder/lib/models/incident.dart
import 'package:shared_models/shared_models.dart';

class ResponderIncident {
  const ResponderIncident._({
    required this.incident,
    required this.responderStatus,
    required String? responderPublicId,
  }) : _responderPublicId = responderPublicId;

  final Incident incident;
  final ResponseStatus? responderStatus;
  final String? _responderPublicId;

  String get publicId => incident.publicId;
  String get title => incident.title;
  String get location => incident.location;
  String get teamPublicId => incident.teamPublicId;
  DateTime? get created => incident.created;
  String get notes => incident.notes;
  String? get priority => incident.priority;
  bool get active => incident.active;
  List<ResponseRecord> get responses => incident.responses;

  ResponseStatus? get status => responderStatus;

  factory ResponderIncident.fromIncident(
    Incident incident, {
    String? responderPublicId,
  }) {
    ResponseStatus? mine;
    if (responderPublicId != null) {
      for (final response in incident.responses) {
        if (response.userPublicId == responderPublicId) {
          mine = response.status;
          break;
        }
      }
    }
    return ResponderIncident._(
      incident: incident,
      responderStatus: mine,
      responderPublicId: responderPublicId,
    );
  }

  factory ResponderIncident.fromJson(
    Map<String, dynamic> json, {
    String? responderPublicId,
  }) =>
      ResponderIncident.fromIncident(
        Incident.fromJson(json),
        responderPublicId: responderPublicId,
      );

  ResponderIncident withResponderResponse(ResponseRecord newResponse) {
    final updated = <ResponseRecord>[];
    var replaced = false;
    for (final existing in incident.responses) {
      if (existing.userPublicId == newResponse.userPublicId) {
        updated.add(newResponse);
        replaced = true;
      } else {
        updated.add(existing);
      }
    }
    if (!replaced) {
      updated.add(newResponse);
    }

    return ResponderIncident.fromIncident(
      incident.copyWith(responses: updated),
      responderPublicId: _responderPublicId,
    );
  }
}
